File: StriverA2Z/C++/3_ArraysQuestion/3.2_Medium/two_sum.cpp
#include<bits/stdc++.h>

using namespace std;

void twoSumMap(int arr[], int target) {
    map<int, int> mpp;
    bool flag = true;
    for (int i = 0; i < (sizeof(arr) / sizeof(arr[0])); i++) {
        int rem = target - arr[i];
        if (mpp.find(rem) != mpp.end()) {
            flag = false;
            cout << "[" << mpp[rem] << "," << i << "]" << endl;
            break;
        }
        mpp[arr[i]] = i;
    }
    if (flag) cout << "[-1,-1]";
}

void twoSumSorted(int arr[], int target) {
    bool flag = true;
    int i = 0;
    int j = (sizeof(arr) / sizeof(arr[0])) - 1;
    while (i < j) {
        int sum = arr[i] + arr[j];
        if (sum == target) {
            flag = false;
            cout << "YES";
            break;
        } else if (sum > target) {
            j--;
        } else {
            i++;
        }
    }
    if (flag) cout << "NO";
}

int main() {
    int arr[] = {2, 7, 11, 14};
    int target = 9;

    twoSumMap(arr, target);
    twoSumSorted(arr, target);
}

File: StriverA2Z/JavaScript&TypeScript/src/Arrays/Medium/twoSum.ts
function twoSumVarientOne(arr: number[], target: number) {
	let flag = true;
	let mpp: Map<number, number> = new Map();
	for (let i = 0; i < arr.length; i++) {
		let rem = target - arr[i];
		if (mpp.has(rem)) {
			flag = false;
			console.log(`[${mpp.get(rem)},${i}]`)
			break;
		}
		mpp.set(arr[i], i);
	}
	if (flag) console.log("[-1,-1]")
}

twoSumVarientOne([2, 6, 5, 8, 11], 14);

function twoSumVarientTwo(arr: number[], target: number) {
	let i = 0;
	let j = arr.length - 1;
	let flag = true;
	while (i < j) {
		let sum = arr[i] + arr[j];
		if (sum === target) {
			console.log("YES");
			flag = false;
			break;
		} else if (sum > target) {
			j--;
		} else {
			i++;
		}
	}
	if (flag) console.log("NO");
}

twoSumVarientTwo([2, 6, 5, 8, 11], 14);
